### ISSUE TITLE
Inconsistent Route Matching in NestJS Based on Route Declaration Order

### ISSUE THREAD

**Issue by brandwork-jb** [nestjs/nest#13104]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current behavior

NestJS exhibits inconsistent behavior in route matching when handling routes with similar structures. Specifically, the routes '/api/v2/test/:type/:id' and '/api/v2/test/:id/token' are resolved differently based on their declaration order in the controller. When the second API endpoint is called, it incorrectly triggers the first endpoint's logic unless the route declaration order is reversed.

The request to /api/v2/test/:id/token is incorrectly processed by the 'getChildsByParentTypeAndId' method when this route is declared before 'getChildTokenByChildId' in the controller. This issue is resolved when the route order is reversed.

### Minimum reproduction code

https://github.com/brandwork-jb/nestjs-routing-inconsistency

### Steps to reproduce

1. Send a request to '/api/v2/test/asdf/token'
2. Observe that the console logs a message from getChildsByParentTypeAndId, incorrectly indicating that this route is being accessed.

Correct Behavior After Reversing Declaration Order:
When the declaration order of the routes is reversed, such that the route with a single parameter (:id/token) is declared first, the route matching works as expected. The request to '/api/v2/test/asdf/token' correctly triggers 'getChildTokenByChildId'.

### Expected behavior

The request to '/api/v2/test/:id/token' should be correctly handled by the 'getChildTokenByChildId' method, regardless of the routes order in the controller.

This inconsistent behavior depending on the order of route declarations could lead to confusing and unpredictable results in applications. A consistent and order-independent route matching mechanism would significantly improve the robustness of NestJS's routing system.

### Package

- [ ] I don't know. Or some 3rd-party package
- [X] <code>@nestjs/common</code>
- [ ] <code>@nestjs/core</code>
- [ ] <code>@nestjs/microservices</code>
- [ ] <code>@nestjs/platform-express</code>
- [ ] <code>@nestjs/platform-fastify</code>
- [ ] <code>@nestjs/platform-socket.io</code>
- [ ] <code>@nestjs/platform-ws</code>
- [ ] <code>@nestjs/testing</code>
- [ ] <code>@nestjs/websockets</code>
- [ ] Other (see below)

### NestJS version

10.3.1

### Packages versions

```
    "@nestjs/apollo": "^12.0.11",
    "@nestjs/axios": "^3.0.1",
    "@nestjs/bull": "^10.0.1",
    "@nestjs/common": "^10.3.1",
    "@nestjs/config": "^3.1.1",
    "@nestjs/core": "^10.3.1",
    "@nestjs/devtools-integration": "^0.1.5",
    "@nestjs/elasticsearch": "^10.0.1",
    "@nestjs/event-emitter": "^2.0.3",
    "@nestjs/graphql": "^12.0.11",
    "@nestjs/platform-express": "^10.3.1",
    "@nestjs/platform-socket.io": "^10.3.1",
    "@nestjs/schedule": "^4.0.0",
    "@nestjs/swagger": "^7.2.0",
    "@nestjs/terminus": "^10.2.0",
    "@nestjs/throttler": "^5.1.1",
    "@nestjs/typeorm": "^10.0.1",
    "@nestjs/websockets": "^10.3.1",
```

### Node.js version

20.10.0

### In which operating systems have you tested?

- [X] macOS
- [ ] Windows
- [ ] Linux

**Comment by brandwork-jb** [comment]
@micalevisk 
I appreciate the feedback and the provided documentation link. My concern isn't with the order of declaration per se, but rather with how different routes seem to conflict with each other despite having a clear distinction. Specifically, the route with a fixed segment (/:id/token) is being overshadowed by the dynamic segment route (/:type/:id), even when "token" is not a variable but a literal segment in the path.

This behavior seems counterintuitive since we usually expect the routing logic to distinguish between dynamic and static segments, making both /api/v2/test/someType/someId and /api/v2/test/someId/token distinct and independently accessible.

Could we consider this as an enhancement to the routing mechanism, enabling it to differentiate between literal and variable segments more effectively, without relying on the declaration order?

Thank you for your time and consideration.

**Comment by micalevisk** [comment]
that's the behavior of the underlying http lib you're using. If you're using Express, it's aligned with Express. Not sure how to handle it better without making our own router, which would turn to be pretty inflexible as Nest is supposed to work with any http lib/framework out there AFIAK

**Comment by soryy708** [comment]
@micalevisk if we don't build it into NestJS, we shouldn't [ostrich algorithm](https://en.wikipedia.org/wiki/Ostrich_algorithm) it either. What can users do on top of NestJS to avoid this foot-gun? How can we direct them to the right thing through documentation?

**Comment by soryy708** [comment]
> The docs are covering this already, aren't they?

@micalevisk Can you please provide a hyperlink to where in the docs this is currently covered?

**Comment by micalevisk** [comment]
https://docs.nestjs.com/controllers#route-parameters

**Comment by soryy708** [comment]
Related:

- https://github.com/nestjs/nest/issues/1667
- https://github.com/nestjs/nest/issues/5931

**Comment by soryy708** [comment]
I see that the docs have:

> HINT
> Routes with parameters should be declared after any static paths. This prevents the parameterized paths from intercepting traffic destined for the static paths.

Now I understand how this relates to this issue, but a very short hint can be hard to follow. Wdyt about adding a code example to the docs, to aid comprehension, and also draw more attention to this foot-gun?

Also, if I dive into a legacy NestJS project, how do I find violations of this rule?

**Comment by micalevisk** [comment]
Yeah, the docs can be improved. Feel free to open a PR with what you think would be better. 

Not sure what do you mean by legacy projects as this has been working like that since the beginning AFIAK. But it wasn't documented before.

We would have to use/make some 3rd-party tool to find such wrong routes. Perhaps this VSC extension could help: https://marketplace.visualstudio.com/items?itemName=hmydgz.nestjs-route-jumper

Maybe that could be suggest to Kamil to add in their audit feature of Nestjs Devtools, idk.

**Comment by agams92** [comment]
For anyone stumbling upon this problem - Fastify adapter is handling these routes as expected, regardless to route declaration order.

<img width="437" height="299" alt="Image" src="https://github.com/user-attachments/assets/98b54b9d-efe6-4f35-8b9a-2e85373fa865" />

<img width="438" height="359" alt="Image" src="https://github.com/user-attachments/assets/dfe68dcc-b08c-4cc6-bc42-e1acadea3db7" />

<img width="397" height="353" alt="Image" src="https://github.com/user-attachments/assets/e18a5852-abbd-485f-9395-8d4e0b79c246" />

<img width="417" height="353" alt="Image" src="https://github.com/user-attachments/assets/09f1b76c-d747-4422-8ebb-d89f0d2bc2e0" />
